<div class="alert alert-block alert-info" style="margin-top: 20px">
    <a href="https://cocl.us/corsera_da0101en_notebook_top">
         <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/Images/TopAd.png" width="750" align="center">
    </a>
</div>

<h1 align=center><font size=5>Data Science and Machine Learning Capstone Project</font></h1>
<h2 align=center><font size=3; color=blue> Copyright &#169; Mustafizur Rahman</font></h2>

<p><b>In this worksheet I will prepare PLUTO data for all Boroughs</b></p>

<h4>Import pandas</h4> 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt

<p><b>Reading Pluto data sets (Bronnx, Brooklyn, Manhatton, Queens and Staten Island)from Computer storage folder.</b></p>

In [2]:
filenameBX = 'E:\\1.IPCOS BACKUP\\3. MMR_PSNL\\2020 Data Science and Analytics Courses\\3. Python Data Science Certificate\\5. Data Science and Machine Learning Capstone Project\\PLUTO_for_WEB\\BX_18v1.csv'
filenameBK = 'E:\\1.IPCOS BACKUP\\3. MMR_PSNL\\2020 Data Science and Analytics Courses\\3. Python Data Science Certificate\\5. Data Science and Machine Learning Capstone Project\\PLUTO_for_WEB\\BK_18v1.csv'
filenameMN = 'E:\\1.IPCOS BACKUP\\3. MMR_PSNL\\2020 Data Science and Analytics Courses\\3. Python Data Science Certificate\\5. Data Science and Machine Learning Capstone Project\\PLUTO_for_WEB\\MN_18v1.csv'
filenameQN = 'E:\\1.IPCOS BACKUP\\3. MMR_PSNL\\2020 Data Science and Analytics Courses\\3. Python Data Science Certificate\\5. Data Science and Machine Learning Capstone Project\\PLUTO_for_WEB\\QN_18v1.csv'
filenameSI = 'E:\\1.IPCOS BACKUP\\3. MMR_PSNL\\2020 Data Science and Analytics Courses\\3. Python Data Science Certificate\\5. Data Science and Machine Learning Capstone Project\\PLUTO_for_WEB\\SI_18v1.csv'

Use the Pandas method <b>read_csv()</b> to load the data from local drive.

In [50]:
# #For Bronnx Dataset
# df_BX_PLUTO = pd.read_csv(filenameBX,low_memory=False)
# print(df_BX_PLUTO.shape)
# df_BX_PLUTO.head(10)
#***************************************************
# #For Brooklyn Dataset
# df_BK_PLUTO = pd.read_csv(filenameBK,low_memory=False)
# print(df_BK_PLUTO.shape)
# df_BK_PLUTO.head(10)
#***************************************************
# #For Manhatton Dataset
# df_MN_PLUTO = pd.read_csv(filenameMN,low_memory=False)
# print(df_MN_PLUTO.shape)
# df_MN_PLUTO.head(10)
#***************************************************
# #For Queens Dataset
# df_QN_PLUTO = pd.read_csv(filenameQN,low_memory=False)
# print(df_QN_PLUTO.shape)
# df_QN_PLUTO.head(10)
#***************************************************
#For Staten Island Dataset
df_SI_PLUTO = pd.read_csv(filenameSI,low_memory=False)
print(df_SI_PLUTO.shape)
df_SI_PLUTO.head(10)

(124493, 87)


Borough  Block  Lot   CD  CT2010  CB2010  SchoolDist  Council  ZipCode  \
0      SI      1   10  501     3.0  2000.0        31.0     49.0  10301.0   
1      SI      1   17  501     3.0     NaN        31.0     49.0  10301.0   
2      SI      1   18  501     3.0     NaN        31.0     49.0  10301.0   
3      SI      1   19  501     3.0     NaN        31.0     49.0  10301.0   
4      SI      1   21  501     3.0     NaN        31.0     49.0  10301.0   
5      SI      1   30  501     3.0     NaN        31.0     49.0  10301.0   
6      SI      1   51  501     3.0  2000.0        31.0     49.0  10301.0   
7      SI      1   55  501     3.0  2000.0        31.0     49.0  10301.0   
8      SI      1   60  501     3.0  2000.0        31.0     49.0  10301.0   
9      SI      1   65  501     3.0  2000.0        31.0     49.0  10301.0   

  FireComp  ...  ZMCode  Sanborn   TaxMap  EDesigNum        APPBBL  \
0     L078  ...     NaN  501 017  50101.0        NaN  0.000000e+00   
1     L078  ...     NaN  501 017  50101.0        NaN  0.000000e+00   
2     L078  ...     NaN  501 017  50101.0        NaN  0.000000e+00   
3     L078  ...     NaN  501 017  50101.0        NaN  0.000000e+00   
4     L078  ...     NaN  501 017  50101.0        NaN  0.000000e+00   
5     L078  ...     NaN  501 017  50101.0        NaN  0.000000e+00   
6     L078  ...     NaN  501 017  50101.0        NaN  5.000010e+09   
7     L078  ...     NaN  501 018  50101.0        NaN  0.000000e+00   
8     L078  ...     NaN  501 018  50101.0        NaN  5.000010e+09   
9     L078  ...     NaN  501 018  50101.0        NaN  5.000010e+09   

      APPDate PLUTOMapID FIRM07_FLAG PFIRM15_FLAG Version  
0         NaN          1         NaN          NaN    18V1  
1         NaN          1         NaN          NaN    18V1  
2         NaN          1         NaN          NaN    18V1  
3         NaN          1         NaN          NaN    18V1  
4         NaN          1         NaN          NaN    18V1  
5         NaN          1         NaN          NaN    18V1  
6  05/02/1996          1         NaN          NaN    18V1  
7         NaN          1         NaN          NaN    18V1  
8  12/26/2013          1         NaN          NaN    18V1  
9  12/26/2013          1         1.0          1.0    18V1  

[10 rows x 87 columns]

<b>Check data Types and Cardinality</b>

In [13]:
df_BX_PLUTO.apply(pd.Series.nunique)

Borough            1
Block           3032
Lot             1960
CD                17
CT2010           339
                ... 
APPDate         1672
PLUTOMapID         4
FIRM07_FLAG        1
PFIRM15_FLAG       1
Version            1
Length: 87, dtype: int64

Create Dataframe with necessary columns and save it for meging with 311 System Data. Data
Columns are:Address, BldgArea, BldgDepth, BuiltFAR, CommFAR,
FacilFAR, Lot, LotArea, LotDepth, NumBldgs, NumFloors, OfficeceArea, ResArea, ResidFAR,
RetailArea, YearBuilt, YearAlter1, ZipCode.

In [51]:
# #For Bronnx Dataset
# df_BX=df_BX_PLUTO[['Borough','Address', 'BldgArea', 'BldgDepth', 'BuiltFAR', 'CommFAR','FacilFAR', 'Lot', 'LotArea', 'LotDepth', 'NumBldgs', 'NumFloors', 'OfficeArea', 'ResArea', 'ResidFAR',
# 'RetailArea', 'YearBuilt', 'YearAlter1', 'ZipCode']]
# df_BX.sort_values(by='YearBuilt',axis=0, ascending=True, kind='mergesort')
# df_BX.head(10)
#***************************************************
# #For Brooklyn Dataset
# df_BK=df_BK_PLUTO[['Borough','Address', 'BldgArea', 'BldgDepth', 'BuiltFAR', 'CommFAR','FacilFAR', 'Lot', 'LotArea', 'LotDepth', 'NumBldgs', 'NumFloors', 'OfficeArea', 'ResArea', 'ResidFAR',
# 'RetailArea', 'YearBuilt', 'YearAlter1', 'ZipCode']]
# df_BK.sort_values(by='YearBuilt',axis=0, ascending=True, kind='mergesort')
# df_BK.head(10)
#***************************************************
# #For Manhatton Dataset
# df_MN=df_MN_PLUTO[['Borough','Address', 'BldgArea', 'BldgDepth', 'BuiltFAR', 'CommFAR','FacilFAR', 'Lot', 'LotArea', 'LotDepth', 'NumBldgs', 'NumFloors', 'OfficeArea', 'ResArea', 'ResidFAR',
# 'RetailArea', 'YearBuilt', 'YearAlter1', 'ZipCode']]
# df_MN.sort_values(by='YearBuilt',axis=0, ascending=True, kind='mergesort')
# df_MN.head(10)
#***************************************************
# #For Queens Dataset
# df_QN=df_QN_PLUTO[['Borough','Address', 'BldgArea', 'BldgDepth', 'BuiltFAR', 'CommFAR','FacilFAR', 'Lot', 'LotArea', 'LotDepth', 'NumBldgs', 'NumFloors', 'OfficeArea', 'ResArea', 'ResidFAR',
# 'RetailArea', 'YearBuilt', 'YearAlter1', 'ZipCode']]
# df_QN.sort_values(by='YearBuilt',axis=0, ascending=True, kind='mergesort')
# df_QN.head(10)
#***************************************************
#For Staten Island Dataset
df_SI=df_SI_PLUTO[['Borough','Address', 'BldgArea', 'BldgDepth', 'BuiltFAR', 'CommFAR','FacilFAR', 'Lot', 'LotArea', 'LotDepth', 'NumBldgs', 'NumFloors', 'OfficeArea', 'ResArea', 'ResidFAR',
'RetailArea', 'YearBuilt', 'YearAlter1', 'ZipCode']]
df_SI.sort_values(by='YearBuilt',axis=0, ascending=True, kind='mergesort')
df_SI.head(10)

Borough            Address  BldgArea  BldgDepth  BuiltFAR  CommFAR  \
0      SI     139 BAY STREET     18965       50.0      3.32      3.4   
1      SI         BAY STREET         0        0.0      0.00      3.4   
2      SI         BAY STREET         0        0.0      0.00      3.4   
3      SI         BAY STREET         0        0.0      0.00      3.4   
4      SI         BAY STREET         0        0.0      0.00      3.4   
5      SI         BAY STREET         0        0.0      0.00      3.4   
6      SI      65 BAY STREET     10600       41.0      2.30      3.4   
7      SI      45 BAY STREET     34800      100.0      2.45      3.4   
8      SI       5 BAY STREET     34941       41.0      0.10      3.4   
9      SI  150 THE PROMENADE      2700       61.0      0.02      3.4   

   FacilFAR  Lot  LotArea  LotDepth  NumBldgs  NumFloors  OfficeArea  ResArea  \
0       4.8   10     5716      84.0         1        4.0           0        0   
1       4.8   17      780      35.0         0        0.0           0        0   
2       4.8   18      520      25.0         0        0.0           0        0   
3       4.8   19      278      15.0         0        0.0           0        0   
4       4.8   21     6302      82.0         0        0.0           0        0   
5       4.8   30        0      80.0         0        0.0           0        0   
6       4.8   51     4600      70.0         1        3.0        9400        0   
7       4.8   55    14200     234.0         1        2.0       34800        0   
8       4.8   60   334565     234.0         4        4.0           0        0   
9       4.8   65   121810       0.0         2        3.0           0        0   

   ResidFAR  RetailArea  YearBuilt  YearAlter1  ZipCode  
0      2.43        6965       1884        2000  10301.0  
1      2.43           0          0           0  10301.0  
2      2.43           0          0           0  10301.0  
3      2.43           0          0           0  10301.0  
4      2.43           0          0           0  10301.0  
5      2.43           0          0           0  10301.0  
6      2.43           0       1923        1991  10301.0  
7      2.43           0       1932           0  10301.0  
8      2.43           0       1931           0  10301.0  
9      2.43           0       1931           0  10301.0

Identify_missing_values

<h4>Evaluating for Missing Data</h4>

The missing values are converted to Python's default. We use Python's built-in functions to identify these missing values. There are two methods to detect missing data:
<ol>
    <li><b>.isnull()</b></li>
    <li><b>.notnull()</b></li>
</ol>
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [52]:
# #For Bronnx Dataset
# BXmissing_data = df_BX.isnull()
# BXmissing_data.head(10)
#***************************************************
#For Brooklyn Dataset
# BKmissing_data = df_BK.isnull()
# BKmissing_data.head(10)
#***************************************************
# #For Manhatton Dataset
# MNmissing_data = df_MN.isnull()
# MNmissing_data.head(10)
#***************************************************
# #For Queens Dataset
# QNmissing_data = df_QN.isnull()
# QNmissing_data.head(10)
#***************************************************
#For Staten Island Dataset
SImissing_data = df_SI.isnull()
SImissing_data.head(10)

Borough  Address  BldgArea  BldgDepth  BuiltFAR  CommFAR  FacilFAR    Lot  \
0    False    False     False      False     False    False     False  False   
1    False    False     False      False     False    False     False  False   
2    False    False     False      False     False    False     False  False   
3    False    False     False      False     False    False     False  False   
4    False    False     False      False     False    False     False  False   
5    False    False     False      False     False    False     False  False   
6    False    False     False      False     False    False     False  False   
7    False    False     False      False     False    False     False  False   
8    False    False     False      False     False    False     False  False   
9    False    False     False      False     False    False     False  False   

   LotArea  LotDepth  NumBldgs  NumFloors  OfficeArea  ResArea  ResidFAR  \
0    False     False     False      False       False    False     False   
1    False     False     False      False       False    False     False   
2    False     False     False      False       False    False     False   
3    False     False     False      False       False    False     False   
4    False     False     False      False       False    False     False   
5    False     False     False      False       False    False     False   
6    False     False     False      False       False    False     False   
7    False     False     False      False       False    False     False   
8    False     False     False      False       False    False     False   
9    False     False     False      False       False    False     False   

   RetailArea  YearBuilt  YearAlter1  ZipCode  
0       False      False       False    False  
1       False      False       False    False  
2       False      False       False    False  
3       False      False       False    False  
4       False      False       False    False  
5       False      False       False    False  
6       False      False       False    False  
7       False      False       False    False  
8       False      False       False    False  
9       False      False       False    False

<b>"True" stands for missing value</b>, while "False" stands for not missing value.

<h4>Count missing values in each column</h4>
<p>
Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value, "False"  means the value is present in the dataset.  In the body of the for loop the method  ".value_counts()"  counts the number of "True" values. 
</p>

In [7]:
# #For Bronnx Dataset
# for column in BXmissing_data.columns.values.tolist():
#     print(column)
#     print (BXmissing_data[column].value_counts())
#     print("") 
#***************************************************
# #For Brooklyn Dataset
# for column in BKmissing_data.columns.values.tolist():
#     print(column)
#     print (BKmissing_data[column].value_counts())
#     print("")
#***************************************************
# #For Manhatton Dataset
# for column in MNmissing_data.columns.values.tolist():
#     print(column)
#     print (MNmissing_data[column].value_counts())
#     print("")
#***************************************************
# #For Queens Dataset
# for column in QNmissing_data.columns.values.tolist():
#     print(column)
#     print (QNmissing_data[column].value_counts())
#     print("")
#***************************************************
# #For Staten Island Dataset
# for column in SImissing_data.columns.values.tolist():
#     print(column)
#     print (SImissing_data[column].value_counts())
#     print("")

Borough
False    89854
Name: Borough, dtype: int64

Address
False    89785
True        69
Name: Address, dtype: int64

BldgArea
False    89854
Name: BldgArea, dtype: int64

BldgDepth
False    89854
Name: BldgDepth, dtype: int64

BuiltFAR
False    89854
Name: BuiltFAR, dtype: int64

CommFAR
False    89854
Name: CommFAR, dtype: int64

FacilFAR
False    89854
Name: FacilFAR, dtype: int64

Lot
False    89854
Name: Lot, dtype: int64

LotArea
False    89854
Name: LotArea, dtype: int64

LotDepth
False    89854
Name: LotDepth, dtype: int64

NumBldgs
False    89854
Name: NumBldgs, dtype: int64

NumFloors
False    89854
Name: NumFloors, dtype: int64

OfficeArea
False    89854
Name: OfficeArea, dtype: int64

ResArea
False    89854
Name: ResArea, dtype: int64

ResidFAR
False    89854
Name: ResidFAR, dtype: int64

RetailArea
False    89854
Name: RetailArea, dtype: int64

YearBuilt
False    89854
Name: YearBuilt, dtype: int64

YearAlter1
False    89854
Name: YearAlter1, dtype: int64

ZipCode
False  

Based on the summary above, <b>Bronx dataset has 89525 rows</b> of data, two columns containing missing data:
<ol>
    <li>"Address": 69 missing data</li>
    <li>"ZipCode": 329 missing data</li>
</ol>
Based on the summary above, <b>Brooklyn dataset has 277316 rows</b> of data, two columns containing missing data:
<ol>
    <li>"Address": 183 missing data</li>
    <li>"ZipCode": 635 missing data</li>
</ol>
Based on the summary above, <b>Manhatton dataset has 42966 rows</b> of data, two columns containing missing data:
<ol>
    <li>"Address": 49 missing data</li>
    <li>"ZipCode": 149 missing data</li>
</ol>
Based on the summary above, <b>Queens dataset has 324583 rows</b> of data, two columns containing missing data:
<ol>
    <li>"Address": 174 missing data</li>
    <li>"ZipCode": 1265 missing data</li>
</ol>
Based on the summary above, <b>Saten Island dataset has 124493 rows</b> of data, two columns containing missing data:
<ol>
    <li>"Address": 37 missing data</li>
    <li>"ZipCode": 1776 missing data</li>
</ol>

<h3 id="deal_missing_values">Deal with missing data</h3>
<b>How to deal with missing data?</b>

<ol>
    <li>drop data<br>
        a. drop the whole row<br>
        b. drop the whole column
    </li>
    <li>replace data<br>
        a. replace it by mean<br>
        b. replace it by frequency<br>
        c. replace it based on other functions
    </li>
</ol>

We will delete all rows where address are missing as we will merge PLUTO data with 311 Complaint Data based on common value of Adress

In [53]:
pd.options.mode.chained_assignment = None
# #For Bronx Dataset
# # simply drop whole row with NaN in "Address" column
# df_BX.dropna(subset=["Address","ZipCode"], axis=0, inplace=True)
# # reset index, because we droped rows
# df_BX.reset_index(drop=True, inplace=True)
# df_BX.head()
#***************************************************
# #For Brooklyn Dataset
# df_BK.dropna(subset=["Address","ZipCode"], axis=0, inplace=True)
# # reset index, because we droped rows
# df_BK.reset_index(drop=True, inplace=True)
# df_BK.head()
#***************************************************
# #For Manhatton Dataset
# df_MN.dropna(subset=["Address","ZipCode"], axis=0, inplace=True)
# # reset index, because we droped rows
# df_MN.reset_index(drop=True, inplace=True)
# df_MN.head(10)
# #***************************************************
# #For Queens Dataset
# df_QN.dropna(subset=["Address","ZipCode"], axis=0, inplace=True)
# # reset index, because we droped rows
# df_QN.reset_index(drop=True, inplace=True)
# df_QN.head(10)
#***************************************************
#For Staten Island Dataset
df_SI.dropna(subset=["Address","ZipCode"], axis=0, inplace=True)
# reset index, because we droped rows
df_SI.reset_index(drop=True, inplace=True)
df_SI.head(10)

Borough            Address  BldgArea  BldgDepth  BuiltFAR  CommFAR  \
0      SI     139 BAY STREET     18965       50.0      3.32      3.4   
1      SI         BAY STREET         0        0.0      0.00      3.4   
2      SI         BAY STREET         0        0.0      0.00      3.4   
3      SI         BAY STREET         0        0.0      0.00      3.4   
4      SI         BAY STREET         0        0.0      0.00      3.4   
5      SI         BAY STREET         0        0.0      0.00      3.4   
6      SI      65 BAY STREET     10600       41.0      2.30      3.4   
7      SI      45 BAY STREET     34800      100.0      2.45      3.4   
8      SI       5 BAY STREET     34941       41.0      0.10      3.4   
9      SI  150 THE PROMENADE      2700       61.0      0.02      3.4   

   FacilFAR  Lot  LotArea  LotDepth  NumBldgs  NumFloors  OfficeArea  ResArea  \
0       4.8   10     5716      84.0         1        4.0           0        0   
1       4.8   17      780      35.0         0        0.0           0        0   
2       4.8   18      520      25.0         0        0.0           0        0   
3       4.8   19      278      15.0         0        0.0           0        0   
4       4.8   21     6302      82.0         0        0.0           0        0   
5       4.8   30        0      80.0         0        0.0           0        0   
6       4.8   51     4600      70.0         1        3.0        9400        0   
7       4.8   55    14200     234.0         1        2.0       34800        0   
8       4.8   60   334565     234.0         4        4.0           0        0   
9       4.8   65   121810       0.0         2        3.0           0        0   

   ResidFAR  RetailArea  YearBuilt  YearAlter1  ZipCode  
0      2.43        6965       1884        2000  10301.0  
1      2.43           0          0           0  10301.0  
2      2.43           0          0           0  10301.0  
3      2.43           0          0           0  10301.0  
4      2.43           0          0           0  10301.0  
5      2.43           0          0           0  10301.0  
6      2.43           0       1923        1991  10301.0  
7      2.43           0       1932           0  10301.0  
8      2.43           0       1931           0  10301.0  
9      2.43           0       1931           0  10301.0

<h3 id="correct_data_format">Correct data format</h3>
<b>We are almost there!</b>
<p>The last step in data cleaning is checking and making sure that all data is in the correct format (int, float, text or other).</p>

In Pandas, we use 
<p><b>.dtype()</b> to check the data type</p>
<p><b>.astype()</b> to change the data type</p>

<h4>Lets list the data types for each column</h4>

In [48]:
df_BX.dtypes

Address        object
BldgArea        int64
BldgDepth     float64
BuiltFAR      float64
CommFAR       float64
FacilFAR      float64
Lot             int64
LotArea         int64
LotDepth      float64
NumBldgs        int64
NumFloors     float64
OfficeArea      int64
ResArea         int64
ResidFAR      float64
RetailArea      int64
YearBuilt       int64
YearAlter1      int64
ZipCode       float64
dtype: object

<p><b>As we can see above, all columns are in the correct data type.</b></p> 
<b>Wonderful!</b>
Now, we finally obtain the cleaned dataset with no missing values and all data in its proper format.
Based on YearBuilt and and YearAlter1 I am calculating the age of the building and last Renovation age in years and add those two columns at the end of the dataset.

<h4>Calculating Age and Renovation age of the building</h4>

In [54]:
import datetime as dtm
current_year = dtm.datetime.now().year #get current year
#print(current_year)

# #For Bronnx Dataset
# #Calculating age of the building
# BldAge =np.where(df_BX["YearBuilt"]!=0, current_year-df_BX["YearBuilt"], df_BX['YearBuilt'])
# BldAge=pd.DataFrame(data=BldAge)
# print(BldAge)
# #calculating Renovation age if renovated
# RenvAge=np.where(df_BX["YearAlter1"]!=0, current_year-df_BX["YearAlter1"], df_BX['YearAlter1'])
# RenvAge=pd.DataFrame(data=RenvAge)
# print(RenvAge)
# #***************************************************
# #For Brooklyn Dataset
# BldAge =np.where(df_BK["YearBuilt"]!=0, current_year-df_BK["YearBuilt"], df_BK['YearBuilt'])
# BldAge=pd.DataFrame(data=BldAge)
# print(BldAge)
# #calculating Renovation age if renovated
# RenvAge=np.where(df_BK["YearAlter1"]!=0, current_year-df_BK["YearAlter1"], df_BK['YearAlter1'])
# RenvAge=pd.DataFrame(data=RenvAge)
# print(RenvAge)
#***************************************************
# #For Manhatton Dataset
# BldAge =np.where(df_MN["YearBuilt"]!=0, current_year-df_MN["YearBuilt"], df_MN['YearBuilt'])
# BldAge=pd.DataFrame(data=BldAge)
# print(BldAge)
# #calculating Renovation age if renovated
# RenvAge=np.where(df_MN["YearAlter1"]!=0, current_year-df_MN["YearAlter1"], df_MN['YearAlter1'])
# RenvAge=pd.DataFrame(data=RenvAge)
# print(RenvAge)
#***************************************************
# #For Queens Dataset
# BldAge =np.where(df_QN["YearBuilt"]!=0, current_year-df_QN["YearBuilt"], df_QN['YearBuilt'])
# BldAge=pd.DataFrame(data=BldAge)
# print(BldAge)
# #calculating Renovation age if renovated
# RenvAge=np.where(df_QN["YearAlter1"]!=0, current_year-df_QN["YearAlter1"], df_QN['YearAlter1'])
# #RenvAge = current_year-df_merged_left["YearAlter1"]
# RenvAge=pd.DataFrame(data=RenvAge)
# print(RenvAge)
#***************************************************
#For Staten Island Dataset
BldAge =np.where(df_SI["YearBuilt"]!=0, current_year-df_SI["YearBuilt"], df_SI['YearBuilt'])
BldAge=pd.DataFrame(data=BldAge)
print(BldAge)
#calculating Renovation age if renovated
RenvAge=np.where(df_SI["YearAlter1"]!=0, current_year-df_SI["YearAlter1"], df_SI['YearAlter1'])
RenvAge=pd.DataFrame(data=RenvAge)
print(RenvAge)

          0
0       136
1         0
2         0
3         0
4         0
...     ...
122707   95
122708  105
122709  105
122710   70
122711   75

[122712 rows x 1 columns]
         0
0       20
1        0
2        0
3        0
4        0
...     ..
122707   0
122708   0
122709   0
122710   0
122711   0

[122712 rows x 1 columns]


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<b>Concate building age with main data frame and Rename it as BldAge .</b>
</div>

In [56]:
#For Bronnx Dataset
#Concate building age with main data frame and Rename it as BldAge
# df_BX = pd.concat([df_BX, BldAge], axis=1)
# df_BX.rename(columns={0:'BldAge'}, inplace=True)
# df_BX.head()
# #Concate Renovation age with main data frame and Rename it as RenvAge
# df_BX = pd.concat([df_BX, RenvAge], axis=1)
# df_BX.head()
# df_BX.rename(columns={0:'RenvAge'}, inplace=True)
# df_BX.tail()
#***************************************************
# #For Brooklyn Dataset
# #Concate building age with main data frame and Rename it as BldAge
# df_BK = pd.concat([df_BK, BldAge], axis=1)
# df_BK.rename(columns={0:'BldAge'}, inplace=True)
# df_BK.head()
# #Concate Renovation age with main data frame and Rename it as RenvAge
# df_BK = pd.concat([df_BK, RenvAge], axis=1)
# df_BK.head()
# df_BK.rename(columns={0:'RenvAge'}, inplace=True)
# df_BK.tail()
#***************************************************
# #For Manhatton Dataset
# #Concate building age with main data frame and Rename it as BldAge
# df_MN = pd.concat([df_MN, BldAge], axis=1)
# df_MN.rename(columns={0:'BldAge'}, inplace=True)
# df_MN.head()
# #Concate Renovation age with main data frame and Rename it as RenvAge
# df_MN = pd.concat([df_MN, RenvAge], axis=1)
# df_MN.head()
# df_MN.rename(columns={0:'RenvAge'}, inplace=True)
# df_MN.tail()
# #***************************************************
# #For Queens Dataset
# df_QN = pd.concat([df_QN, BldAge], axis=1)
# df_QN.rename(columns={0:'BldAge'}, inplace=True)
# df_QN.head()
# #Concate Renovation age with main data frame and Rename it as RenvAge
# df_QN = pd.concat([df_QN, RenvAge], axis=1)
# df_QN.head()
# df_QN.rename(columns={0:'RenvAge'}, inplace=True)
# df_QN.tail()
#***************************************************
# #For Staten Island Dataset
# #Concate building age with main data frame and Rename it as BldAge
# df_SI = pd.concat([df_SI, BldAge], axis=1)
# df_SI.rename(columns={0:'BldAge'}, inplace=True)
# df_SI.head()
#Concate Renovation age with main data frame and Rename it as RenvAge
df_SI = pd.concat([df_SI, RenvAge], axis=1)
df_SI.rename(columns={0:'RenvAge'}, inplace=True)
df_SI.tail()

Borough             Address  BldgArea  BldgDepth  BuiltFAR  CommFAR  \
122707      SI  142 BENTLEY STREET      1624       26.0      0.23      0.0   
122708      SI  146 BENTLEY STREET      2057       26.0      0.15      0.0   
122709      SI  150 BENTLEY STREET      1728       32.0      0.16      0.0   
122710      SI  156 BENTLEY STREET      1408       21.0      0.21      0.0   
122711      SI  162 BENTLEY STREET      1768       24.0      0.44      0.0   

        FacilFAR  Lot  LotArea  LotDepth  ...  NumFloors  OfficeArea  ResArea  \
122707       1.0   83     7200    120.00  ...        2.5           0     1624   
122708       1.0   86    14136    228.00  ...        2.5           0     2057   
122709       1.0   89    10664    273.00  ...        2.0           0     1728   
122710       1.0   92     6630    136.17  ...        2.0           0     1408   
122711       1.0   94     4000     80.00  ...        2.0           0     1768   

        ResidFAR  RetailArea  YearBuilt  YearAlter1  ZipCode  BldAge  RenvAge  
122707       0.5           0       1925           0  10307.0      95        0  
122708       0.5           0       1915           0  10307.0     105        0  
122709       0.5           0       1915           0  10307.0     105        0  
122710       0.5           0       1950           0  10307.0      70        0  
122711       0.5           0       1945           0  10307.0      75        0  

[5 rows x 21 columns]

<b>Wonderful!</b>

Now, we have clean and <b>reduced PLUTO dataset</b> to merge withcomplaint type data set. 

<b>DATA NORMALIZATION!</b>
NOrmalization is a process of trnasforming values of several variables into a similar range. Typically scalling variables so that the variables values range from 0 to 1.

If using sag and saga solvers <b>FOR LOGISTIC REGRATION OR ANY OTHER CLASSIFICATION MODEL</b>, make sure the features are on a similar scale. We saw the importance of this above. Andreas Mueller, in private correspondence, also mentioned that he found convergence issues on unscaled data with lbfgs, although it was more robust than sag and saga.
<b>Bottom line: to be safe, scale your data.</b>. In our next model for other Bourough's I will use Scalling feature matrix and try with Linear regression!

<b>For this dataset we need to scale/normalize folloing Variables:(THESE VARIABLES WILL BE USED AS A FEATURE MATRIX FOR MODELING</b> 
BldgArea	BldgDepth	BuiltFAR	CommFAR	FacilFAR	LotArea	LotDepth	NumBldgs	NumFloors	OfficeArea	ResArea	ResidFAR	RetailArea	BldAge and	RenvAge

In [57]:
# #For Bronnx Dataset
# df_BX['BldgArea'] =df_BX['BldgArea']/df_BX['BldgArea'].max()
# df_BX['BldgDepth'] =df_BX['BldgDepth']/df_BX['BldgDepth'].max()
# df_BX['BuiltFAR'] =df_BX['BuiltFAR']/df_BX['BuiltFAR'].max()
# df_BX['CommFAR'] =df_BX['CommFAR']/df_BX['CommFAR'].max()
# df_BX['FacilFAR'] =df_BX['FacilFAR']/df_BX['FacilFAR'].max()
# df_BX['LotArea'] =df_BX['LotArea']/df_BX['LotArea'].max()
# df_BX['LotDepth'] =df_BX['LotDepth']/df_BX['LotDepth'].max()
# df_BX['NumBldgs'] =df_BX['NumBldgs']/df_BX['NumBldgs'].max()
# df_BX['NumFloors'] =df_BX['NumFloors']/df_BX['NumFloors'].max()
# df_BX['OfficeArea'] =df_BX['OfficeArea']/df_BX['OfficeArea'].max()
# df_BX['ResArea'] =df_BX['ResArea']/df_BX['ResArea'].max()
# df_BX['ResidFAR'] =df_BX['ResidFAR']/df_BX['ResidFAR'].max()
# df_BX['RetailArea'] =df_BX['RetailArea']/df_BX['RetailArea'].max()
# df_BX['BldAge'] =df_BX['BldAge']/df_BX['BldAge'].max()
# df_BX['RenvAge'] =df_BX['RenvAge']/df_BX['RenvAge'].max()
# df_BX.tail()
#***************************************************
# #For Brooklyn Dataset
# df_BK['BldgArea'] =df_BK['BldgArea']/df_BK['BldgArea'].max()
# df_BK['BldgDepth'] =df_BK['BldgDepth']/df_BK['BldgDepth'].max()
# df_BK['BuiltFAR'] =df_BK['BuiltFAR']/df_BK['BuiltFAR'].max()
# df_BK['CommFAR'] =df_BK['CommFAR']/df_BK['CommFAR'].max()
# df_BK['FacilFAR'] =df_BK['FacilFAR']/df_BK['FacilFAR'].max()
# df_BK['LotArea'] =df_BK['LotArea']/df_BK['LotArea'].max()
# df_BK['LotDepth'] =df_BX['LotDepth']/df_BK['LotDepth'].max()
# df_BK['NumBldgs'] =df_BK['NumBldgs']/df_BK['NumBldgs'].max()
# df_BK['NumFloors'] =df_BK['NumFloors']/df_BK['NumFloors'].max()
# df_BK['OfficeArea'] =df_BK['OfficeArea']/df_BK['OfficeArea'].max()
# df_BK['ResArea'] =df_BK['ResArea']/df_BK['ResArea'].max()
# df_BK['ResidFAR'] =df_BK['ResidFAR']/df_BK['ResidFAR'].max()
# df_BK['RetailArea'] =df_BK['RetailArea']/df_BK['RetailArea'].max()
# df_BK['BldAge'] =df_BK['BldAge']/df_BK['BldAge'].max()
# df_BK['RenvAge'] =df_BK['RenvAge']/df_BK['RenvAge'].max()
# df_BK.tail()
#***************************************************
# #For Manhatton Dataset
# df_MN['BldgArea'] =df_MN['BldgArea']/df_MN['BldgArea'].max()
# df_MN['BldgDepth'] =df_MN['BldgDepth']/df_MN['BldgDepth'].max()
# df_MN['BuiltFAR'] =df_MN['BuiltFAR']/df_MN['BuiltFAR'].max()
# df_MN['CommFAR'] =df_MN['CommFAR']/df_MN['CommFAR'].max()
# df_MN['FacilFAR'] =df_MN['FacilFAR']/df_MN['FacilFAR'].max()
# df_MN['LotArea'] =df_MN['LotArea']/df_MN['LotArea'].max()
# df_MN['LotDepth'] =df_MN['LotDepth']/df_MN['LotDepth'].max()
# df_MN['NumBldgs'] =df_MN['NumBldgs']/df_MN['NumBldgs'].max()
# df_MN['NumFloors'] =df_MN['NumFloors']/df_MN['NumFloors'].max()
# df_MN['OfficeArea'] =df_MN['OfficeArea']/df_MN['OfficeArea'].max()
# df_MN['ResArea'] =df_MN['ResArea']/df_MN['ResArea'].max()
# df_MN['ResidFAR'] =df_MN['ResidFAR']/df_MN['ResidFAR'].max()
# df_MN['RetailArea'] =df_MN['RetailArea']/df_MN['RetailArea'].max()
# df_MN['BldAge'] =df_MN['BldAge']/df_MN['BldAge'].max()
# df_MN['RenvAge'] =df_MN['RenvAge']/df_MN['RenvAge'].max()
# df_MN.tail()
# #***************************************************
# #For Queens Dataset
#For Manhatton Dataset
# df_QN['BldgArea'] =df_QN['BldgArea']/df_QN['BldgArea'].max()
# df_QN['BldgDepth'] =df_QN['BldgDepth']/df_QN['BldgDepth'].max()
# df_QN['BuiltFAR'] =df_QN['BuiltFAR']/df_QN['BuiltFAR'].max()
# df_QN['CommFAR'] =df_QN['CommFAR']/df_QN['CommFAR'].max()
# df_QN['FacilFAR'] =df_QN['FacilFAR']/df_QN['FacilFAR'].max()
# df_QN['LotArea'] =df_QN['LotArea']/df_QN['LotArea'].max()
# df_QN['LotDepth'] =df_QN['LotDepth']/df_QN['LotDepth'].max()
# df_QN['NumBldgs'] =df_QN['NumBldgs']/df_QN['NumBldgs'].max()
# df_QN['NumFloors'] =df_QN['NumFloors']/df_QN['NumFloors'].max()
# df_QN['OfficeArea'] =df_QN['OfficeArea']/df_QN['OfficeArea'].max()
# df_QN['ResArea'] =df_QN['ResArea']/df_QN['ResArea'].max()
# df_QN['ResidFAR'] =df_QN['ResidFAR']/df_QN['ResidFAR'].max()
# df_QN['RetailArea'] =df_QN['RetailArea']/df_QN['RetailArea'].max()
# df_QN['BldAge'] =df_QN['BldAge']/df_QN['BldAge'].max()
# df_QN['RenvAge'] =df_QN['RenvAge']/df_QN['RenvAge'].max()
# df_QN.tail()
#***************************************************
#For Staten Island Dataset
df_SI['BldgArea'] =df_SI['BldgArea']/df_SI['BldgArea'].max()
df_SI['BldgDepth'] =df_SI['BldgDepth']/df_SI['BldgDepth'].max()
df_SI['BuiltFAR'] =df_SI['BuiltFAR']/df_SI['BuiltFAR'].max()
df_SI['CommFAR'] =df_SI['CommFAR']/df_SI['CommFAR'].max()
df_SI['FacilFAR'] =df_SI['FacilFAR']/df_SI['FacilFAR'].max()
df_SI['LotArea'] =df_SI['LotArea']/df_SI['LotArea'].max()
df_SI['LotDepth'] =df_SI['LotDepth']/df_SI['LotDepth'].max()
df_SI['NumBldgs'] =df_SI['NumBldgs']/df_SI['NumBldgs'].max()
df_SI['NumFloors'] =df_SI['NumFloors']/df_SI['NumFloors'].max()
df_SI['OfficeArea'] =df_SI['OfficeArea']/df_SI['OfficeArea'].max()
df_SI['ResArea'] =df_SI['ResArea']/df_SI['ResArea'].max()
df_SI['ResidFAR'] =df_SI['ResidFAR']/df_SI['ResidFAR'].max()
df_SI['RetailArea'] =df_SI['RetailArea']/df_SI['RetailArea'].max()
df_SI['BldAge'] =df_SI['BldAge']/df_SI['BldAge'].max()
df_SI['RenvAge'] =df_SI['RenvAge']/df_SI['RenvAge'].max()
df_SI.tail()

Borough             Address  BldgArea  BldgDepth  BuiltFAR  CommFAR  \
122707      SI  142 BENTLEY STREET  0.001674   0.005179  0.002597      0.0   
122708      SI  146 BENTLEY STREET  0.002121   0.005179  0.001694      0.0   
122709      SI  150 BENTLEY STREET  0.001781   0.006375  0.001807      0.0   
122710      SI  156 BENTLEY STREET  0.001452   0.004183  0.002372      0.0   
122711      SI  162 BENTLEY STREET  0.001823   0.004781  0.004969      0.0   

        FacilFAR  Lot   LotArea  LotDepth  ...  NumFloors  OfficeArea  \
122707  0.208333   83  0.000226  0.012475  ...   0.021930         0.0   
122708  0.208333   86  0.000444  0.023703  ...   0.021930         0.0   
122709  0.208333   89  0.000335  0.028381  ...   0.017544         0.0   
122710  0.208333   92  0.000208  0.014156  ...   0.017544         0.0   
122711  0.208333   94  0.000126  0.008317  ...   0.017544         0.0   

         ResArea  ResidFAR  RetailArea  YearBuilt  YearAlter1  ZipCode  \
122707  0.002483  0.166667         0.0       1925           0  10307.0   
122708  0.003145  0.166667         0.0       1915           0  10307.0   
122709  0.002642  0.166667         0.0       1915           0  10307.0   
122710  0.002153  0.166667         0.0       1950           0  10307.0   
122711  0.002703  0.166667         0.0       1945           0  10307.0   

          BldAge  RenvAge  
122707  0.267606      0.0  
122708  0.295775      0.0  
122709  0.295775      0.0  
122710  0.197183      0.0  
122711  0.211268      0.0  

[5 rows x 21 columns]

In [58]:
#save the clean dataframe for Merging and Data Analysis as well as Predictive Modeling
#df_BX.to_csv('BX_Clean_PLUTO_df.csv',index=False)
#df_BK.to_csv('BK_Clean_PLUTO_df.csv',index=False)
#df_MN.to_csv('MN_Clean_PLUTO_df.csv',index=False)
#df_QN.to_csv('QN_Clean_PLUTO_df.csv',index=False)
df_SI.to_csv('SI_Clean_PLUTO_df.csv',index=False)

<hr>
<p>Copyright &copy; 2020 Mustafizur Rahman <a href="https://www.linkedin.com/in/dr-mohammad-rahman-2434b313/">LinkedIn</a>.</p>